src/main/asciidoc/index.adoc

```diff
@@ -17,503 +17,801 @@
+* the authorization server: a server issuing tokens to the client after successfully authenticating the resource owner
+and obtaining authorization.
+
+The authorization process of OAuth 2.0 utilizes two authorization server endpoints:
+
+* the authorization endpoint: an endpoint used by the client to obtain authorization from the resource owner;
+* the token endpoint: an endpoint used by the client to exchange an authorization grant for an access token.
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+To access a protected resource, the client MUST pass the access token in the Authorization header of the request using
+the Bearer HTTP authentication scheme (defined in RFC 6750 as part of the HTTP authentication framework).
+====
 
-* Prepare your HTTP message using the JWS.
-** The access tokens in OAuth 2.0 are of type *bearer*, meaning the client needs to pass the token with each request.
-** The followings are the expected HTTP headers:
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+The authorization server MUST be configured so that it returns access tokens that are self-contained in lieu of opaque
+tokens.
+====
 
-[width="100%",cols="22%,19%,17%,42%",]
-|===
-|*HTTP Header* |*Description* |*Required* |*Value*
-|Authorization |Must be set to the access token. |Must |Bearer \{Access Token}
-|Content-Type |Content type of the payload. |Optional a|
-\{JSON body} -> application/jose+json
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+Indeed, opaque tokens MUST be validated at the authorization server's token introspection endpoint for each request,
+which is grossly inefficient.
+====
 
-\{Non JSON body} -> application/jose
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+Self-contained tokens, by contrast, SHOULD be validated locally by the resource server, assuming that there is a
+pre-established trust relationship between the resource server and the authorization server.
+====
 
-|===
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+For RESTful services processing Category 3-4 data, it's RECOMMENDED to validate all access tokens at the authorization
+server's token introspection endpoint (despite the performance penalty).
+====
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+As any other confidential, access tokens MUST be kept confidential in transit and at rest, and only shared among the
+authorization server, the resource servers the access token is valid for (in a micro-services architecture, access
+tokens MAY be propagated between services), and the client to whom the access token is issued.
+====
+
+Self-contained tokens contain a set of claims stored in the header part of an
+https://datatracker.ietf.org/doc/html/rfc7519[RFC 7519] compliant JSON Web Token (JWT; pronounced "jot") object
+represented as either a https://datatracker.ietf.org/doc/html/rfc7515[JSON Web Signature (JWS)] and/or
```

Comment:
Self-contained tokens contain a set of claims stored in the **payload** part of an
 https://datatracker.ietf.org/doc/html/rfc7519[RFC 7519] compliant JSON Web Token (JWT; pronounced "jot") object represented as either a https://datatracker.ietf.org/doc/html/rfc7515[JSON Web Signature (JWS)] and/or
 https://datatracker.ietf.org/doc/html/rfc7516[JSON Web Encryption (JWE)] structure.